src/westmere/ip4.h

```diff
@@ -0,0 +1,210 @@
+  // load, but it is not generally available with SSE 4.1.
+  const __m128i input = _mm_loadu_si128((const __m128i *)ipv4_string);
+  if (ipv4_string_length > 15) {
+    return 0;
+  }
+  // locate dots
+  uint16_t dotmask;
+  {
+    const __m128i dot = _mm_set1_epi8('.');
+    const __m128i t0 = _mm_cmpeq_epi8(input, dot);
+    dotmask = _mm_movemask_epi8(t0);
+    uint16_t mask = 1 << ipv4_string_length;
+    dotmask &= mask - 1;
+    dotmask |= mask;
+  }
+
+  // build a hashcode
+  const uint8_t hashcode = ((6639 * dotmask) >> 13);
+
+  // grab the index of the shuffle mask
+  const uint8_t id = patterns_id[hashcode];
+  if (id >= 81) {
+    return 0;
+  }
+  const uint8_t *pat = &patterns[id][0];
+  const __m128i pattern = _mm_loadu_si128((const __m128i *)pat);
+  // The value of the shuffle mask at a specific index points at the last digit,
+  // we check that it matches the length of the input.
+  const __m128i ascii0 = _mm_set1_epi8('0');
+  const __m128i t0 = input;
+  __m128i t1 = _mm_shuffle_epi8(t0, pattern);
+  // check that leading digits of 2- 3- numbers are not zeros.
+  {
+    const __m128i eq0 = _mm_cmpeq_epi8(t1, ascii0);
+    if (!_mm_testz_si128(eq0, _mm_set_epi8(-1, 0, -1, 0, -1, 0, -1, 0,
+                                           0, 0, 0, 0, 0, 0, 0, 0))) {
+      return 0;
+    }
+  }
+  // replace null values with '0'
+  __m128i t1b = _mm_blendv_epi8(t1, ascii0, pattern);
+
+  // subtract '0'
+  const __m128i t2 = _mm_sub_epi8(t1b, ascii0);
+  // check that everything was in the range '0' to '9'
+  {
+    const __m128i c9 = _mm_set1_epi8('9' - '0');
+    const __m128i t2m = _mm_max_epu8(t2, c9);
+    const __m128i t2me = _mm_cmpeq_epi8(t2m, c9);
+    if (!_mm_test_all_ones(t2me)) {
+      return 0;
+    }
+  }
+  // We do the computation, the Mula way.
+  const __m128i weights =
+      _mm_setr_epi8(1, 10, 1, 10, 1, 10, 1, 10, 100, 0, 100, 0, 100, 0, 100, 0);
+  const __m128i t3 = _mm_maddubs_epi16(t2, weights);
+  const __m128i t4 = _mm_alignr_epi8(t3, t3, 8);
+  const __m128i t5 = _mm_add_epi16(t4, t3);
+  // Test that we don't overflow (over 255)
+  if (!_mm_testz_si128(t5, _mm_set_epi8(0, 0, 0, 0, 0, 0, 0, 0, -1, 0, -1,
+                                        0, -1, 0, -1, 0))) {
+    return 0;
+  }
+  // pack and we are done!
+  const __m128i t6 = _mm_packus_epi16(t5, t5);
+  uint32_t address =  _mm_cvtsi128_si32(t6);
+  memcpy(destination, &address, 4);
+  return 1;
+}
+
+zone_always_inline()
+zone_nonnull_all()
+static inline void parse_ip4(
+  zone_parser_t *parser,
+  const zone_type_info_t *type,
+  const zone_field_info_t *field,
+  zone_token_t *token)
+{
+  if (token->length > INET_ADDRSTRLEN)
+    SEMANTIC_ERROR(parser, "Invalid %s in %s",
+                   field->name.data, type->name.data);
+  // Note that this assumes that reading up to token->data + 16 is safe (i.e., we do not cross a page).
+  if (sse_inet_aton(token->data, token->length, &parser->rdata->octets[parser->rdata->length]) != 1)
```

Comment:
`token->length` will not be available after #64 is merged. Preferably, we account for that at this stage (we can worry about it later too, but your knowledge of the algorithm is still fresh now?). In an attempt to vectorize timestamp deserialization (#22, still WiP), I opted to use the result of the subtract operation to detect where the IP address ends. The maximum string size is 16 bytes (INET_ADDRSTRLEN), we can use that to get the dot mask instead of `ipv4_string_length`(?), then later on use the result of the subtract operation combined with the found dots to determine what should be kept? Then we can verify with the length wether or not the character that follows the address is a valid delimiter. At least, that's what I think is a better approach than keeping the length with the token... (as always, this is open for suggestions/discussion)

Reply:
Taking a closer look at the code (there may be things I did not account for, and I'm still playing with it, but I wanted to share):
```
  const __m128i ascii0_9 = _mm_setr_epi8(
    '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 0, 0, 0, 0, 0, 0);
  const __m128i digits = _mm_cmpeq_epi8(input, _mm_shuffle_epi8(ascii0_9, input));

  // locate dots
  uint16_t dotmask;
  { 
    const __m128i dot = _mm_set1_epi8('.');
    dotmask = (uint16_t)_mm_movemask_epi8(_mm_cmpeq_epi8(input, dot));
    const uint16_t digit_mask = (uint16_t)_mm_movemask_epi8(digits);
    uint16_t mask = ~(digit_mask | dotmask);
    mask &= -mask;
    dotmask &= mask - 1;
    dotmask |= mask;
    // FIXME: testing for weird input can now be done using masks but perhaps we can
    //        catch it when determining hashcode too?
  }
```

We can then do the subtract before the shuffle:
```
  const __m128i t0 = _mm_and_si128(_mm_sub_epi8(input, ascii0), digits);
```

But that does allow us to drop the blend and the range check?

Reply:
> But that does allow us to drop the blend and the range check?
It looks like it does, since it slices the input at the first bad char.

Here is a untested version with different bit twiddling:
```
uint16_t m = digit_mask | dotmask;
m ^= (m + 1); // mask of lowest clear bit and below
dotmask = ~digitmask & m;
// string_length = popcnt or lzcnt   
```